src/codewhisperer/util/codeWhispererSession.ts

```diff
@@ -0,0 +1,68 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { CodewhispererCompletionType } from '../../shared/telemetry/telemetry.gen'
+import { GenerateRecommendationsRequest, ListRecommendationsRequest, Recommendation } from '../client/codewhisperer'
+import { Position } from 'vscode'
+import { CodeWhispererSupplementalContext } from './supplementalContext/supplementalContextUtil'
+
+const performance = globalThis.performance ?? require('perf_hooks').performance
+
+class CodeWhispererSession {
+    static #instance: CodeWhispererSession
+
+    // Per-session states
+    sessionId = ''
+    startPos = new Position(0, 0)
+    requestContext: {
+        request: ListRecommendationsRequest | GenerateRecommendationsRequest
+        supplementalMetadata: Omit<CodeWhispererSupplementalContext, 'supplementalContextItems'> | undefined
+    } = { request: {} as any, supplementalMetadata: {} as any }
+
+    // Various states of recommendations
+    recommendations: Recommendation[] = []
+    suggestionStates = new Map<number, string>()
+    completionTypes = new Map<number, CodewhispererCompletionType>()
+
+    // Some other variables for client component latency
+    fetchCredentialStartTime = 0
+    sdkApiCallStartTime = 0
+    invokeSuggestionStartTime = 0
+
+    public static get instance() {
+        return (this.#instance ??= new CodeWhispererSession())
+    }
+
+    setFetchCredentialStart() {
+        if (this.fetchCredentialStartTime === 0 && this.invokeSuggestionStartTime !== 0) {
+            this.fetchCredentialStartTime = performance.now()
+        }
+    }
+
+    setSdkApiCallStart() {
+        if (this.sdkApiCallStartTime === 0 && this.fetchCredentialStartTime !== 0) {
+            this.sdkApiCallStartTime = performance.now()
+        }
+    }
+
+    setSuggestionState(index: number, value: string) {
+        this.suggestionStates.set(index, value)
+    }
+
+    getSuggestionState(index: number): string | undefined {
+        return this.suggestionStates.get(index)
+    }
+
+    setCompletionType(index: number, recommendation: Recommendation) {
+        const nonBlankLines = recommendation.content.split('\n').filter(line => line.trim() !== '').length
+        this.completionTypes.set(index, nonBlankLines > 1 ? 'Block' : 'Line')
+    }
+
+    getCompletionType(index: number): CodewhispererCompletionType {
+        return this.completionTypes.get(index) || 'Line'
+    }
+}
+
+export const session = CodeWhispererSession.instance
```

Comment:
Does this call `instance()` at moduel-load time? That can lead to strange order-of-operations issues, because the module load order will affect when `instance()` is called. That can be non-obvious to troubleshoot.

Not a blocker for this PR, but as a followup, this could be fixed by converting `session` to a function.